specification/flag-evaluation/flag-evaluation.md

```diff
@@ -0,0 +1,195 @@
+
+```
+
+##### Requirement 1.11
+
+> The `evaluation details` wrapper **MUST** have a `value` field matching the type of the flag value, and containing to the flag value.
+
+##### Condition 1.12
+
+> The language supports generics (or an equivalent feature).
+>
+> ##### Conditional Requirement 1.12.1
+>
+> > The `evaluation details` wrapper **SHOULD** accept a generic argument (or utilize an equivalent language feature) which indicates the type of the wrapped `value` field.
+
+##### Requirement 1.13
+
+> The `evaluation details` wrapper **MUST** have a `flag key` field field of type string, representing the flag key.
+
+##### Requirement 1.14
+
+> The `evaluation details` wrapper **MUST** have an optional `variant` field of type string which represents a space-constrained identifier for a particular flag value.
+
+##### Requirement 1.15
+
+> The `evaluation details` wrapper **MUST** have a `reason` field of type string indicating the semantic reason for the returned flag value. Possible values vary by provider, but may include `"TARGETING_MATCH"`, `"SPLIT"`, `"DISABLED"`, `"DEFAULT"`, `"UNKNOWN"` and `"ERROR"`.
+
+##### Requirement 1.16
+
+> In cases of abnormal execution (network failure, unhandled error, etc) the `reason` field in the `evaluation details` **MUST** be set to `"ERROR"`.
+
+##### Requirement 1.17
+
+> The `evaluation details` wrapper **MUST** have an optional `error code` field of type string, which, in cases of abnormal execution, identifies an error occurred during flag evaluation, with possible values `"PROVIDER_NOT_READY"`, `"FLAG_NOT_FOUND"`, `"PARSE_ERROR"`, `"TYPE_MISMATCH"`, or `"GENERAL"`.
+
+##### Requirement 1.18
+
+> If the `reason` in the `evaluation details` is set to `"ERROR"`, the sibling `error code` field **MUST** be set.
+
+##### Requirement 1.19
+
+> The `evaluation details` wrapper **MUST** have a `executed hooks` field, containing an object which itself contains 4 arrays, identified with keys (`before`, `after`, `error` and `finally`) indicating the hooks run for the respective life-cycle stages.
+
+```
+executedHooks: {
+  before: [],
+  after: [],
+  error: [],
+  finally: []
+}
+```
+
+##### Requirement 1.20
+
+> The value of the elements in the arrays contained in the `executed hooks` object **SHOULD** allow authors to uniquely identify the hooks executed during the flag evaluation lifecycle. This could be done with a mandatory string field defined on the `hook` interface, or by using the fully qualified class name in languages where such concepts exist.
+
+##### Requirement 1.21
+
+> The `evaluation options` object **MUST** have a `hooks` field which can contain hooks to be executed for the respective flag evaluation.
+
+see [hooks](./hooks.md)
+
+##### Requirement 1.22
+
+> No methods, functions, or operations on the client should ever throw exceptions, or otherwise abnormally end. Flag evaluation calls must always return the `default value` in the event of abnormal execution.
```

Comment:
I would carve out an exception for configuration and setup. If something is incorrectly setup, then failing on startup is a reasonable choice and swallowing that error might be confusing.

Reply:
I specifically mentioned "client" here, because most configuration is done by the point you have an instance of one, with the exception of adding hooks to the client... I think it's a fair point though, to say this rule really applies to flag evaluation only.

Reply:
One thing came up for me:  This is still true for the case where something in the `finally()` hook errors out, right? If that's the case, I think it will make the logic quite odd for handling those exceptions, making it not much different than `after`

Reply:
I think the `finally` or `afterAll` hook is pretty similar functionally to the `after` hook. In my mind, the big difference is that the `after` hook would only be executed if flag evaluation proceeds normally, while the `finally` hook executes in both normal, and abnormal `error` execution. What do you think?